README.md
![Duckling Logo](https://github.com/facebook/duckling/raw/master/logo.png)

# Duckling [![Build Status](https://travis-ci.org/facebook/duckling.svg?branch=master)](https://travis-ci.org/facebook/duckling)
Duckling is a Haskell library that parses text into structured data.

```
"the first Tuesday of October"
=> {"value":"2017-10-03T00:00:00.000-07:00","grain":"day"}
```

## Requirements
A Haskell environment is required. We recommend using
[stack](https://haskell-lang.org/get-started).

On macOS you'll need to install PCRE development headers.
The easiest way to do that is with [Homebrew](https://brew.sh/):
```
brew install pcre
```
If that doesn't help, try running `brew doctor` and fix
the issues it finds.

## Quickstart
To compile and run the binary:
```
$ stack build
$ stack exec duckling-example-exe
```
The first time you run it, it will download all required packages.

This runs a basic HTTP server. Example request:
```
$ curl -XPOST http://0.0.0.0:8000/parse --data 'locale=en_GB&text=tomorrow at eight'
```
In the example application, all dimensions are enabled by default. Provide the parameter `dims` to specify which ones you want. Examples:
```
Identify credit card numbers only:
$ curl -XPOST http://0.0.0.0:8000/parse --data 'locale=en_US&text="4111-1111-1111-1111"&dims="["credit-card-number"]"'
If you want multiple dimensions, comma-separate them in the array:
$ curl -XPOST http://0.0.0.0:8000/parse --data 'locale=en_US&text="3 cups of sugar"&dims="["quantity","numeral"]"'
```

See `exe/ExampleMain.hs` for an example on how to integrate Duckling in your
project.
If your backend doesn't run Haskell or if you don't want to spin your own Duckling server, you can directly use [wit.ai](https://wit.ai)'s built-in entities.

## Supported dimensions
Duckling supports many languages, but most don't support all dimensions yet
(**we need your help!**).
Please look into [this directory](https://github.com/facebook/duckling/blob/master/Duckling/Dimensions) for language-specific support.

| Dimension | Example input | Example value output
| --------- | ------------- | --------------------
| `AmountOfMoney` | "42€" | `{"value":42,"type":"value","unit":"EUR"}`
| `CreditCardNumber` | "4111-1111-1111-1111" | `{"value":"4111111111111111","issuer":"visa"}`
| `Distance` | "6 miles" | `{"value":6,"type":"value","unit":"mile"}`
| `Duration` | "3 mins" | `{"value":3,"minute":3,"unit":"minute","normalized":{"value":180,"unit":"second"}}`
| `Email` | "duckling-team@fb.com" | `{"value":"duckling-team@fb.com"}`
| `Numeral` | "eighty eight" | `{"value":88,"type":"value"}`
| `Ordinal` | "33rd" | `{"value":33,"type":"value"}`
| `PhoneNumber` | "+1 (650) 123-4567" | `{"value":"(+1) 6501234567"}`
| `Quantity` | "3 cups of sugar" | `{"value":3,"type":"value","product":"sugar","unit":"cup"}`
| `Temperature` | "80F" | `{"value":80,"type":"value","unit":"fahrenheit"}`
| `Time` | "today at 9am" | `{"values":[{"value":"2016-12-14T09:00:00.000-08:00","grain":"hour","type":"value"}],"value":"2016-12-14T09:00:00.000-08:00","grain":"hour","type":"value"}`
| `Url` | "https://api.wit.ai/message?q=hi" | `{"value":"https://api.wit.ai/message?q=hi","domain":"api.wit.ai"}`
| `Volume` | "4 gallons" | `{"value":4,"type":"value","unit":"gallon"}`

[Custom dimensions](https://github.com/facebook/duckling/blob/master/exe/CustomDimensionExample.hs) are also supported.

## Extending Duckling
To regenerate the classifiers and run the test suite:
```
$ stack build :duckling-regen-exe && stack exec duckling-regen-exe && stack test
```

It's important to regenerate the classifiers after updating the code and before
running the test suite.

To extend Duckling's support for a dimension in a given language, typically 4
files need to be updated:
* `Duckling/<Dimension>/<Lang>/Rules.hs`
* `Duckling/<Dimension>/<Lang>/Corpus.hs`
* `Duckling/Dimensions/<Lang>.hs` (if not already present in `Duckling/Dimensions/Common.hs`)
* `Duckling/Rules/<Lang>.hs`

To add a new language:
* Make sure that the language code used follows the [ISO-639-1 standard](https://en.wikipedia.org/wiki/List_of_ISO_639-1_codes).
* The first dimension to implement is `Numeral`.
* Follow [this example](https://github.com/facebook/duckling/commit/24d3f199768be970149412c95b1c1bf5d76f8240).

To add a new locale:
* There should be a need for diverging rules between the locale and the language.
* Make sure that the locale code is a valid [ISO3166 alpha2 country code](https://www.iso.org/obp/ui/#search/code/).
* Follow [this example](https://github.com/facebook/duckling/commit/1ab5f447d2635fe6d48887a501d333a52adff5b9).

Rules have a name, a pattern and a production.
Patterns are used to perform character-level matching (regexes on input) and
concept-level matching (predicates on tokens).
Productions are arbitrary functions that take a list of tokens and return a new
token.

The corpus (resp. negative corpus) is a list of examples that should (resp.
shouldn't) parse. The reference time for the corpus is Tuesday Feb 12, 2013 at
4:30am.

`Duckling.Debug` provides a few debugging tools:
```
$ stack repl --no-load
> :l Duckling.Debug
> debug (makeLocale EN $ Just US) "in two minutes" [Seal Time]
in|within|after <duration> (in two minutes)
-- regex (in)
-- <integer> <unit-of-duration> (two minutes)
-- -- integer (0..19) (two)
-- -- -- regex (two)
-- -- minute (grain) (minutes)
-- -- -- regex (minutes)
[Entity {dim = "time", body = "in two minutes", value = RVal Time (TimeValue (SimpleValue (InstantValue {vValue = 2013-02-12 04:32:00 -0200, vGrain = Second})) [SimpleValue (InstantValue {vValue = 2013-02-12 04:32:00 -0200, vGrain = Second})] Nothing), start = 0, end = 14}]
```

## License
Duckling is BSD-licensed.


CHANGELOG.md
# Changelog for [`duckling` package](https://github.com/facebook/duckling)

## 0.2.X.X

### Core
* Make `isRangeValid` behave differently based on lang
* Add custom `isRangeValid` implementation for ZH

### Rulesets
* CA (Catalan)
  * AmountOfMoney: **new!**
  * Distance: **new!**
  * Duration: **new!**
  * Temperature: **new!**
  * Time: **new!**
  * Volume: **new!**

* EN (English)
  * Time: Allow latent match for \<part-of-day\> \<latent-time-of-day\>
  * Time: Avoid parsing phrases like 'two five' as times
  * Time: Add support to parse spelled-out times like 'five-thirty'

* ES (Spanish)
  * Numeral: Fix parsing for hundreds phrases like `doscientos tres`
  * Numeral: Add support for thousands and millions

* RU (Russian)
  * Duration: Diminutives for minutes and hours
  * Duration: quarters of an hour
  * Duration: added 'сутки' (24-hour period)
  * Duration: use coarsest reasonable grain for \<positive-non-integer\> \<time-grain\>

### Server

## 0.2.0.0

### Core
  * Bump versions on dependencies
  * ExampleMain: add support for setting the port
  * Support for GHC >= 8.8.x && GHC <= 9.0.1
  * Relicense to BSD-3
  * Probabilistic layer bug fix
  * Remove dependency on Data.Some
  * Depend on regex-pcre-builtin on Windows
  * Update cabal-version to 2.2
  * Migrate to GitHub CI over Travis
  * Some attempts at getting smaller parse trees (less exploitable)

### Rulesets

* Common
  * Correct CDT TimeZone offset  CDT is (-300), not 540.
  * Fixed BST and IST offsets. BST is +1 (60) and IST is +5:30 (330)
  * Make unicode output in tests sane by not relying on `Show`
  * Add type=value to JSON response for Email, PhoneNumber, and Url, for consistency
  * CreditCardNumber: **new!** (duckling can now recognise credit cards)
  * Email: Add DE (German) + IS (Icelandic) spelled out email
  * Numeral: Don't accept dashes ('-') as token separators
  * Time: Add Karva Chauth holiday
  * Time: Add Vaisakhi holiday
  * Time: Add daylight savings start/endtimes to holidays
  * Time: Add Purim and Shushan Purim (Jewish holidays)
  * Time: Add Guru Gobind Singh Jayanti holiday
  * Time: Extend support for Ramadan and Eid al-Fitr from 1950 to 2050
  * Time: Add support for Parsi New Year
  * Time: Add support for Dayananda Saraswati Jayanti holiday
  * Time: Extend support for Mawlid from 1950 to 1998
  * Time: Add Rabindra Jayanti holiday
  * Time: Add Guru Ravidass Jayanti holiday
  * Time: Extend support for Eid al-Adha from 1950 to 2000
  * Time: Add Krishna Janmashtami holiday
  * Time: Add Mahavir Jayanti holiday
  * Time: Add Maha Shivaratri holiday
  * Time: add Ugadi holiday
  * Volume: interval support

* AF (Afrikaans)
  * Numeral: **new!**

* AR (Arabic)
  * Add a new locale: EG (Egyptian)
  * AmountOfMoney: Add more variants of EGP
  * Numeral: Add support for numerals written in Arabic script
  * Numeral: Support decimals and comma-separated integers
  * PhoneNumber: **new!**
  * Time: Add periodic times
  * Time: Add rule for a week ago

* BG (Bulgarian)
  * Time: **new!**

* BN (Bengali)
  * Numeral: **new!**

* CA (Catalan)
  * Numeral: **new!**
  * Ordinal: **new!**

* DA (Danish)
  * Ordinal: Add support for larger spelled-out ordinals
  * Time: recognise abbreviation 'kl'

* DE (German)
  * Distance: **new!**
  * Duration: Add more common durations
  * Numeral: Fix typo on "fünfzehn"(15)
  * Time: Fix wrong parsing of YYYY-MM-DD dates
  * Time: Add rule for 'the day before yesterday'
  * Time: Fix a bug for "fünfter"¬
  * Time: support for "am ersten Dezember" to " am einunddreißigsten Dezember"¬
  * Time: now recognise "der fünfte Dezember"¬
  * Time: Add support for (pre-)computed holidays
  * Time: Don't parse 'so'
  * Time: improve approximations
  * Time: Add rule for week after next
  * Time: Add rule for montag in n wochen
  * Volume: **new!**

* EN (English)
  * AmountOfMoney: Add support for subunits of dollars, e.g. nickels, dimes, quarters, as well as the number of coins
  * AmountOfMoney: Add support for latent
  * AmountOfMoney: Add support for lakh and crore
  * AmountOfMoney: correct regex to match three letter currency codes beginning with C
  * AmountOfMoney: Make ruleInterval{Max,Min} symmetric
  * AmountOfMoney: Support abbreviations of lakh and crore
  * AmountOfMoney: Add UAH currency
  * AmountOfMoney: Add HKD
  * AmountOfMoney: dollar and \<amount-of-money\>
  * AmountOfMoney: dollar and a half
  * AmountOfMoney: intersection for \<amount-of-money\> and a half
  * AmountOfMoney: extend interval support
  * AmountOfMoney: Add more variants of EGP
  * Distance: Support composite distances
  * Distance: one meter and \<distance\>
  * Distance: \<distance\> meters and \<distance\>
  * Distance: Add interval support
  * Duration: Fix composite durations without delimiters.
              It previously only worked with
              commas/'and'-separated tokens
  * Duration: Support more words in `ruleDurationNumeralMore`
  * Duration: \<integer\> and a half minutes
  * Duration: support for \<integer\> hour and \<integer\>
  * Duration: parse Xm as X minutes and X.Y hrs as X.Y hours
  * Duration: Leverage TimeGrain for x.y hours
  * Duration: Support composite duration
  * Numeral: Fix ambiguous parses when both ruleNegative and ruleMultiply apply
  * Numeral: Extend fraction rule
  * Numeral: Add mixed fraction rules
  * Numeral: Add prefix of 10/100/10_000 rules
  * Quantity: Support 'k.g' and 'k.g.' for kilograms
  * Temperature: Support temperatures with decimal values
  * Time: Fix palm sunday regex
  * Time: Fix latent time of days like "ten thirty"
  * Time: Don't parse time of days above 12 with meridiem
          like "13 am"
  * Time: Add rule to match \<time\> + \<timezone\>
  * Time: Add "all week"/"rest of the week" rules
  * Time: Parse time intervals like: 2015-03-28 17:00:00/2015-03-28 21:00:00
  * Time: Add rules to handle "DD/MM/YYYY" or "DD MM YYYY"
  * Time: Add support for YYYYQQ and YYQQ expressions, like
          '2018Q4' and '18Q4'

Duckling/Types.hs
-- Copyright (c) 2016-present, Facebook, Inc.
-- All rights reserved.
--
-- This source code is licensed under the BSD-style license found in the
-- LICENSE file in the root directory of this source tree.


{-# LANGUAGE AllowAmbiguousTypes #-}
{-# LANGUAGE DeriveAnyClass #-}
{-# LANGUAGE DeriveGeneric #-}
{-# LANGUAGE ExistentialQuantification #-}
{-# LANGUAGE FlexibleContexts #-}
{-# LANGUAGE FlexibleInstances #-}
{-# LANGUAGE GADTs #-}
{-# LANGUAGE LambdaCase #-}
{-# LANGUAGE NoRebindableSyntax #-}
{-# LANGUAGE OverloadedStrings #-}
{-# LANGUAGE RankNTypes #-}
{-# LANGUAGE ScopedTypeVariables #-}
{-# LANGUAGE StandaloneDeriving #-}
{-# LANGUAGE TypeFamilies #-}
{-# LANGUAGE TypeOperators #-}
{-# LANGUAGE NamedFieldPuns #-}

module Duckling.Types where

import Control.DeepSeq
import Data.Aeson
import Data.GADT.Compare
import Data.GADT.Show
import Data.Hashable
import Data.HashMap.Strict (HashMap)
import Data.HashSet (HashSet)
import Data.List (intersperse, sortOn)
import Data.Maybe
import Data.Text (Text, toLower, unpack)
import Data.Typeable ((:~:)(Refl), eqT, Typeable)
import GHC.Generics
import Prelude
import TextShow (TextShow(..))
import qualified Data.ByteString.Lazy as LB
import qualified Data.HashMap.Strict as HashMap
import qualified Data.Text as TT
import qualified Data.Text.Encoding as Text
import qualified Text.Regex.Base as R
import qualified Text.Regex.PCRE as PCRE
import qualified TextShow as TS

import Duckling.AmountOfMoney.Types (AmountOfMoneyData)
import Duckling.CreditCardNumber.Types (CreditCardNumberData)
import Duckling.Distance.Types (DistanceData)
import Duckling.Duration.Types (DurationData)
import Duckling.Email.Types (EmailData)
import Duckling.Locale
import Duckling.Numeral.Types (NumeralData)
import Duckling.Ordinal.Types (OrdinalData)
import Duckling.PhoneNumber.Types (PhoneNumberData)
import Duckling.Quantity.Types (QuantityData)
import Duckling.Regex.Types
import Duckling.Resolve
import Duckling.Temperature.Types (TemperatureData)
import Duckling.Time.Types (TimeData)
import Duckling.TimeGrain.Types (Grain)
import Duckling.Url.Types (UrlData)
import Duckling.Volume.Types (VolumeData)

-- -----------------------------------------------------------------
-- Token

data Token = forall a . (Resolve a, Eq a, Hashable a, Show a, NFData a) =>
  Token (Dimension a) a

deriving instance Show Token
instance Eq Token where
  Token d1 v1 == Token d2 v2 = case geq d1 d2 of
    Just Refl -> v1 == v2
    Nothing   -> False

instance Hashable Token where
  hashWithSalt s (Token d v) = hashWithSalt s (d, v)

instance NFData Token where
  rnf (Token _ v) = rnf v

-- -----------------------------------------------------------------
-- Seal

data Seal s where
  Seal :: s a -> Seal s

instance GEq s => Eq (Seal s) where
  Seal x == Seal y =
    defaultEq x y

instance GShow s => Show (Seal s) where
  showsPrec p (Seal s)
    = showParen (p > 10) (showString "Seal " . gshowsPrec 11 s)

withSeal :: Seal s -> (forall t. s t -> r) -> r
withSeal (Seal x) f = f x

-- -----------------------------------------------------------------
-- Dimension

class (Show a, Typeable a, Typeable (DimensionData  a)) =>
    CustomDimension a where
  type DimensionData a
  dimRules :: a -> [Rule]
  dimLangRules :: Lang -> a -> [Rule]
  dimLocaleRules :: Region -> a -> [Rule]
  dimDependents :: a -> HashSet (Seal Dimension)

-- | GADT for differentiating between dimensions
-- Each dimension should have its own constructor and provide the data structure
-- for its parsed data
data Dimension a where
  RegexMatch :: Dimension GroupMatch
  AmountOfMoney :: Dimension AmountOfMoneyData
  CreditCardNumber :: Dimension CreditCardNumberData
  Distance :: Dimension DistanceData
  Duration :: Dimension DurationData
  Email :: Dimension EmailData
  Numeral :: Dimension NumeralData
  Ordinal :: Dimension OrdinalData
  PhoneNumber :: Dimension PhoneNumberData
  Quantity :: Dimension QuantityData
  Temperature :: Dimension TemperatureData
  Time :: Dimension TimeData
  TimeGrain :: Dimension Grain
  Url :: Dimension UrlData
  Volume :: Dimension VolumeData
  CustomDimension :: CustomDimension a => a -> Dimension (DimensionData a)

-- Show
instance Show (Dimension a) where
  show RegexMatch = "RegexMatch"
  show CreditCardNumber = "CreditCardNumber"
  show Distance = "Distance"
  show Duration = "Duration"
  show Email = "Email"
  show AmountOfMoney = "AmountOfMoney"
  show Numeral = "Numeral"
  show Ordinal = "Ordinal"
  show PhoneNumber = "PhoneNumber"
  show Quantity = "Quantity"
  show Temperature = "Temperature"
  show Time = "Time"
  show TimeGrain = "TimeGrain"
  show Url = "Url"
  show Volume = "Volume"
  show (CustomDimension d) = show d
instance GShow Dimension where gshowsPrec = showsPrec

-- TextShow
instance TextShow (Dimension a) where
  showb d = TS.fromString $ show d
instance TextShow (Seal Dimension) where
  showb (Seal d) = showb d

-- Hashable
instance Hashable (Seal Dimension) where
  hashWithSalt s (Seal a) = hashWithSalt s a
instance Hashable (Dimension a) where
  hashWithSalt s RegexMatch          = hashWithSalt s (0::Int)
  hashWithSalt s Distance            = hashWithSalt s (1::Int)
  hashWithSalt s Duration            = hashWithSalt s (2::Int)
  hashWithSalt s Email               = hashWithSalt s (3::Int)
  hashWithSalt s AmountOfMoney       = hashWithSalt s (4::Int)
  hashWithSalt s Numeral             = hashWithSalt s (5::Int)
  hashWithSalt s Ordinal             = hashWithSalt s (6::Int)
  hashWithSalt s PhoneNumber         = hashWithSalt s (7::Int)
  hashWithSalt s Quantity            = hashWithSalt s (8::Int)
  hashWithSalt s Temperature         = hashWithSalt s (9::Int)
  hashWithSalt s Time                = hashWithSalt s (10::Int)
  hashWithSalt s TimeGrain           = hashWithSalt s (11::Int)
  hashWithSalt s Url                 = hashWithSalt s (12::Int)
  hashWithSalt s Volume              = hashWithSalt s (13::Int)
  hashWithSalt s (CustomDimension _) = hashWithSalt s (14::Int)
  hashWithSalt s CreditCardNumber    = hashWithSalt s (15::Int)

instance GEq Dimension where
  geq RegexMatch RegexMatch = Just Refl
  geq RegexMatch _ = Nothing
  geq CreditCardNumber CreditCardNumber = Just Refl
  geq CreditCardNumber _ = Nothing
  geq Distance Distance = Just Refl
  geq Distance _ = Nothing
  geq Duration Duration = Just Refl
  geq Duration _ = Nothing
  geq Email Email = Just Refl
  geq Email _ = Nothing
  geq AmountOfMoney AmountOfMoney = Just Refl
  geq AmountOfMoney _ = Nothing
  geq Numeral Numeral = Just Refl
  geq Numeral _ = Nothing
  geq Ordinal Ordinal = Just Refl
  geq Ordinal _ = Nothing
  geq PhoneNumber PhoneNumber = Just Refl
  geq PhoneNumber _ = Nothing
  geq Quantity Quantity = Just Refl
  geq Quantity _ = Nothing
  geq Temperature Temperature = Just Refl
  geq Temperature _ = Nothing
  geq Time Time = Just Refl
  geq Time _ = Nothing
  geq TimeGrain TimeGrain = Just Refl
  geq TimeGrain _ = Nothing
  geq Url Url = Just Refl
  geq Url _ = Nothing
  geq Volume Volume = Just Refl
  geq Volume _ = Nothing
  geq (CustomDimension (_ :: a)) (CustomDimension (_ :: b))
    | Just Refl <- eqT :: Maybe (a :~: b) = Just Refl
  geq (CustomDimension _) _ = Nothing

isDimension :: Dimension a -> Token -> Bool
isDimension dim (Token dim' _) = isJust $ geq dim dim'

data ResolvedVal
  = forall a . ( Resolve a, Eq (ResolvedValue a)
               , Show (ResolvedValue a)
               , ToJSON (ResolvedValue a)) =>
    RVal (Dimension a) (ResolvedValue a)

deriving instance Show ResolvedVal

instance Eq ResolvedVal where
  RVal d1 v1 == RVal d2 v2
    | Just Refl <- geq d1 d2 = v1 == v2
    | otherwise = False

data Node = Node
  { nodeRange :: Range
  , token     :: Token
  , children  :: [Node]
  , rule      :: Maybe Text
  } deriving (Eq, Generic, Hashable, Show, NFData)

data ResolvedToken = Resolved
  { range :: Range
  , node :: Node
  , rval :: ResolvedVal
  , isLatent :: Bool
  } deriving (Eq, Show)

instance Ord ResolvedToken where
  compare (Resolved range1 _ (RVal _ v1) latent1)
          (Resolved range2 _ (RVal _ v2) latent2) =
    case compare range1 range2 of
      EQ -> case compare (toJText v1) (toJText v2) of
        EQ -> compare latent1 latent2
        z -> z
      z  -> z


-- |A Candidate represents a potential match going into the ranker
data Candidate = Candidate
    ResolvedToken -- ^ The actual resolved token we are considering
    Double -- ^ naive Bayes log-likelihood - sum of LL of all rules used
    Bool -- ^ Does the ResolvedToken's dimension match the caller's request?
  deriving (Eq, Show)

instance Ord Candidate where
  compare (Candidate Resolved{range = Range s1 e1, node = Node{token = Token d1 _}} score1 t1)
          (Candidate Resolved{range = Range s2 e2, node = Node{token = tok2}} score2 t2)
    | isDimension d1 tok2 = case starts of
        EQ -> case ends of
          EQ -> compare score1 score2
          z -> z
        LT -> case ends of
          LT -> EQ
          _ -> GT
        GT -> case ends of
          GT -> EQ
          _ -> LT
    | t1 == t2 = compRange
    | t1 && compRange == GT = GT
    | t2 && compRange == LT = LT
    | otherwise = EQ
      where
        starts = compare s1 s2
        ends = compare e1 e2
        -- a > b if a recovers b
        compRange = case starts of
          EQ -> ends
          LT -> case ends of
            LT -> EQ
            _  -> GT
          GT -> case ends of
            GT -> EQ
            _  -> LT

data Range = Range Int Int
  deriving (Eq, Ord, Generic, Hashable, Show, NFData)

type Production = [Token] -> Maybe Token
type Predicate = Token -> Bool
data PatternItem = Regex PCRE.Regex | Predicate Predicate

type Pattern = [PatternItem]

data Rule = Rule
  { name :: Text
  , pattern :: Pattern
  , prod :: Production
  }

instance Show Rule where
  show Rule{ name } = show name

data Entity = Entity
  { dim    :: Text
  , body   :: Text
  , value  :: ResolvedVal
  , start  :: Int
  , end    :: Int
  , latent :: Bool
  , enode  :: Node
  } deriving (Eq, Generic, Show)

instance ToJSON Entity where
  toJSON ent@Entity{value = RVal _ val} = object
    [ "dim"    .= dim ent
    , "body"   .= body ent
    , "value"  .= val
    , "start"  .= start ent
    , "end"    .= end ent
    , "latent" .= latent ent
    ]

toJText :: ToJSON x => x -> Text
toJText = Text.decodeUtf8 . LB.toStrict . encode

-- -----------------------------------------------------------------
-- Predicates helpers

regex :: String -> PatternItem
regex = Regex . R.makeRegexOpts compOpts execOpts
  where
    compOpts = PCRE.defaultCompOpt + PCRE.compCaseless + PCRE.compUTF8
    execOpts = PCRE.defaultExecOpt

dimension :: Typeable a => Dimension a -> PatternItem
dimension d = Predicate $ isDimension d

-- -----------------------------------------------------------------
-- Rule Construction helpers

singleStringLookupRule :: HashMap Text a -> Text -> (a -> Maybe Token) -> Rule
singleStringLookupRule hashMap name production = Rule
  { name
  , pattern = [ regex $ unpack regexString ]
  , prod = \case
      (Token RegexMatch (GroupMatch (match:_)):_) ->
        HashMap.lookup (toLower match) hashMap >>= production
      _ -> Nothing
  }
  where
    regexString =
      "(" <> mconcat
      (intersperse "|" $ sortOn (negate . TT.length) $ HashMap.keys hashMap)
      <> ")"
